### ISSUE TITLE
Read- and Write- buffering stream classes are not provided/inaccessible

### ISSUE THREAD

**Issue by PeterTruchly** [Microsoft/vs-streamjsonrpc#52]
While it is possible to implement own transport protocol by extending DelimitedMessageHandler, the class ReadBufferingStream is made internal to the library.
Moreover, write buffering class would be welcome in order to increase efficiency when using network streams (like SslStream).

Writing and flushing each message separately, like in HeaderDelimitedMessageHandler.cs lines 240-242 is suboptimal, especially for large quantity of small messages.
~~~ C#
            // Transmit the content itself.
            await this.SendingStream.WriteAsync(contentBytes, 0, contentBytes.Length, cancellationToken).ConfigureAwait(false);
            await this.SendingStream.FlushAsync().ConfigureAwait(false);
~~~
_Note:_ Simply using BufferedStream does not solve the whole problem, as flushing it per each call will introduce the same inefficiency and not-flushing it right away brings the question where and when to do it.

**Comment by AArnott** [comment]
I'm not sure what you're suggesting here for write buffering. A call to `JsonRpc.InvokeAsync` must actually submit the call, so any buffering must be flushed after every call. And the way `HeaderDelimitedMessageHandler` transmits the message doesn't look particularly inefficient to me. The only enhancement that occurs to me at the moment is to recycle the header buffer we use so that in your case of many messages we don't reallocate so many buffers. Is there another optimization you can think of?

As for the `ReadBufferingStream`, we might expose that if you needed it. But if you're targeting a framework that offers `BufferedStream` I would recommend you use that for reading. [It isn't available on netstandard1.1](https://docs.microsoft.com/en-us/dotnet/api/system.io.bufferedstream?view=netstandard-1.1) which is why we had to implement it ourselves.

**Comment by PeterTruchly** [comment]
I would say 'any buffering must be flushed after every call' only if there is nothing else to do.

Ok, I will explain my view on this. Just imagine that for some reason we would like to stream calls between two nodes, what is of particular interest here is throughput and latency. Now, in the stream (e.g. network stream) every call/message occupies a certain amount of consecutive bytes. You are right about submitting each call, I am only questioning the necessity of flushing each call separately. From my point of view - submitting the call means it must be transmitted as soon as possible using the stream.

1. If there is no ongoing operation on the stream, the message is written. 'Flush' operation is queued.
2. If there is any ongoing operation, 'write' is queued, all 'flush' operations are removed from queue and one 'flush' op. is placed at the end of a queue.

This way, in case of many messages/operations going through the same stream at the same time, flush is postponed until there is nothing more to write.

**Comment by AArnott** [comment]
That makes sense. The way it's written we have no way to know whether more requests will be written in order to skip a Flush call. Also, some of those writes may be canceled before they're even written. So suppose a Write call skipped the call to Flush because another writer was about to start. But then that writer is canceled. Now no one calls flush.

I imagine we could solve it, but it would be rather complex and hard to exhaustively test. I wonder if this is a problem better solved by the application using this library by passing it a `BufferedStream` to write to, and then the application can make requests as much as it wants and then Flush the BufferedStream once. A wrapper around `BufferedStream` could actually be what's passed to JsonRpc so that when it calls Flush on it, it's ignored, but the app who has the inner stream can call it to force it.

**Comment by IlyaBiryukov** [comment]
I think only DelimitedMessageHandler knows when another writer has expressed interest in writing to the stream. The underlying stream doesn't have this context. Also there is no facility on the Stream class to let it know that someone else is about to write something to it soon.

It looks like an extensibility point on the delimited message handler may enable the proposed improvement.

**Comment by PeterTruchly** [comment]
Well, perhaps an example is worth more than just talk. Lets consider two operations on a stream: 1. writing a message, which costs 1ms (lets call it 'M') and 2. flushing of stream which costs 10ms (lets call it 'F'). What is being done now, for each message, there is a flush operation, effectively doing M,F, costing us 11ms (within this given example). 
Lets assume now, we would like to send a message M every 3ms...
What happens now, we do M,F,M,F,M,F which gives us throughput of 90 messages/s.
What is possible to do (as mentioned in Your posts) to do M, M, M, M, M, F (at least 5 messages in a row, then flush once) which gives throughput 333messages/s (in theory, delivery times are not evenly spaced, but average time per message is 3ms).
This could be done either explicitly (Your suggestion), -OR- internally (my recommendation).
Please consider the following:
Stream is 'idle', M arrives which takes 1ms to write, there is operation F scheduled and executed right away (there is no next message yet). While F is executing, after 2ms... the next 'M' arrives, after 5ms another 'M', after 8ms the third 'M'. After 10ms the F operation is done, but now 3 messages are queued. Writing of first M takes 1ms and another M arrives by that time. So this is still 3 messages in a queue. Writing MMM takes another 3ms, 1 more M arrives... 1ms later and everything is written. Now there is nothing to do, so we should execute 'F' right away... 
In other words, there is an autobalancing mechanism possible.
- frequent messaging causes less stream flushes but increased throughput
- infrequent messaging causes more stream flushes but immediate (best effort) delivery is still guaranteed

**Comment by AArnott** [comment]
Thanks for writing that out, @PeterTruchly. I'm concerned about the typical scenario of:
Message M is written. Time T elapses before we even flush the message because we're waiting to see if more messages are coming, then we finally flush in time T. So the total elapsed time for a single message grows by T. In an extremely common scenario, the caller is waiting for a response from the server, so we'd be making these exchanges unnecessarily longer. That's unacceptable. 

But consider this: if given your scenario of sending 3 M's in a row, while we currently call FlushAsync after each M, we aren't _blocking_ on each flush. So in fact if you send 3 M's in a row, or even 10 of them, you may end up only flushing twice and sending a bunch in a batch. Do you have evidence of the worst case scenario you described actually happens?

**Comment by PeterTruchly** [comment]
> while we currently call FlushAsync after each M, we aren't blocking on each flush

All operations on a stream are serialized at the end, no matter what you call. What is more important, **every stream operation takes time to complete, while the stream is blocked and cannot process any other operation**.

> Time T elapses before we even flush the message because we're waiting to see if more messages are coming

I am not suggesting to wait, by 'best effort' I mean: we should send messages as fast as we can without any arbitrary wait period.

> Do you have evidence of the worst case scenario you described actually happens?

I measured and found more than tenfold difference. Just for the record, I am using stream from TcpClient wrapped by SslStream wrapped by BufferedStream.

> So in fact if you send 3 M's in a row, or even 10 of them, you may end up only flushing twice and sending a bunch in a batch.

It is not as asynchronous and independent as you describe, (`WriteCoreAsync` carries out write and flush in sequence):

~~~
using (await this.sendingSemaphore.EnterAsync(cts.Token).ConfigureAwait(false))
{
     await this.WriteCoreAsync(content, contentEncoding, cts.Token).ConfigureAwait(false);
}
~~~

while:

~~~
private readonly AsyncSemaphore sendingSemaphore = new AsyncSemaphore(1);
~~~

**Comment by AArnott** [comment]
> It is not as asynchronous and independent as you describe, (WriteCoreAsync carries out write and flush in sequence)

I see. I suspect we can optimize that by only invoking `WriteCoreAsync` within the semaphore but without awaiting it, but I'll have to check. If we await outside the semaphore, then my hope would be that many folks can send messages independently, and although a Stream's FlushAsync method is invoked, it would still allow folks to write to the stream, but I guess we'll have to test that.

**Comment by AArnott** [comment]
@PeterTruchly I think I've found several improvements I can make to this area. Can you share your test app that you used to confirm the problem so I can use it to verify my fixes?
